data/modules/CargoRun/CargoRun.lua

```diff
@@ -0,0 +1,1138 @@
+		wholesaler	= wholesaler,
+		pickup		= pickup,
+		introtext	= introtext,
+		dist		= dist,
+		due		= due,
+		cargo		= cargo,
+		branch		= branch,
+		cargotype	= cargotype,
+		risk		= risk,
+		urgency		= urgency,
+		reward		= reward,
+		faceseed	= Engine.rand:Integer(),
+	}
+
+	n = getNumberOf("ADTEXT_" .. missiontype)
+	local adtext
+	if n >= 1 then
+		adtext = l["ADTEXT_" .. missiontype .. "_" .. Engine.rand:Integer(1, n)]
+	else
+		adtext = l["ADTEXT_" .. Engine.rand:Integer(1, getNumberOf("ADTEXT"))]
+	end
+	ad.desc = string.interp(adtext, {
+		system	 = nearbysystem.name,
+		cash	 = Format.Money(ad.reward),
+		starport = ad.location:GetSystemBody().name,
+	})
+
+	local ref = station:AddAdvert({
+		description = ad.desc,
+		icon        = ad.urgency >=  0.8 and "haul_fast" or "haul",
+		onChat      = onChat,
+		onDelete    = onDelete,
+		isEnabled   = isEnabled })
+	ads[ref] = ad
+
+	return ad
+end
+
+local onCreateBB = function (station)
+	local num = Engine.rand:Integer(0, math.ceil(Game.system.population))
+	for i = 1,num do
+		makeAdvert(station)
+	end
+end
+
+local onUpdateBB = function (station)
+	for ref,ad in pairs(ads) do
+		if ad.localdelivery then
+			if ad.due < Game.time + 2*60*60*24 then -- two day timeout for locals
+				ad.station:RemoveAdvert(ref)
+			end
+		else
+			if ad.due < Game.time + 5*60*60*24 then -- five day timeout for inter-system
+				ad.station:RemoveAdvert(ref)
+			end
+		end
+	end
+	if Engine.rand:Integer(12*60*60) < 60*60 then -- roughly once every twelve hours
+		makeAdvert(station)
+	end
+end
+
+local pirate_ships = {}
+local pirate_gripes_time
+local pirate_switch_target
+local escort_ships = {}
+local escort_chatter_time
+local escort_switch_target
+
+local onEnterSystem = function (player)
+	if (not player:IsPlayer()) then return end
+
+	local syspath = Game.system.path
+
+	for ref,mission in pairs(missions) do
+		if mission.status == "ACTIVE" and ((mission.location:IsSameSystem(syspath) and not mission.pickup) or (mission.domicile:IsSameSystem(syspath) and mission.cargo_picked_up)) then
+			local risk = mission.risk
+			local pirates = 0
+
+			local riskmargin = Engine.rand:Number(-0.3,0.3) -- Add some random luck
+			if risk >= (1 + riskmargin) then pirates = 3
+			elseif risk >= (0.7 + riskmargin) then pirates = 2
+			elseif risk >= (0.5 + riskmargin) then pirates = 1
+			end
+
+			-- if there is some risk and still no pirates, flip a tricoin
+			if pirates < 1 and risk >= 0.2 and Engine.rand:Integer(2) == 1 then pirates = 1 end
+
+			-- XXX hull mass is a bad way to determine suitability for role
+			local shipdefs = utils.build_array(utils.filter(function (k,def) return def.tag == 'SHIP'
+				and def.hyperdriveClass > 0 and def.hullMass <= 400 end, pairs(ShipDef)))
+			if #shipdefs == 0 then return end
+
+			local pirate
+
+			while pirates > 0 do
+				pirates = pirates - 1
+
+				if Engine.rand:Number(1) <= risk then
+					local shipdef = shipdefs[Engine.rand:Integer(1,#shipdefs)]
```

Comment:
Mpfff... We really should introduce a standard way to spawn random ships. Being the one that broke the hyperdriveClass API and had to go and fix all these snippets, I can tell you there is a *lot* of cargo-cult coding going on :-)